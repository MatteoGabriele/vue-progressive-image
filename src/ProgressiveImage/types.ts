export type ProgressiveImageProps = {
  src: string;
  placeholderSrc?: string;
  fallbackSrc?: string;
  alt?: string;
  title?: string;
  customClass?: string;
  blur?: number;
  lazyPlaceholder?: boolean;
  objectCover?: boolean;
  delay?: number;
};

export type ProgressiveImageGlobalProps = Omit<
  ProgressiveImageProps,
  "src" | "objectCover | title | alt"
>;
