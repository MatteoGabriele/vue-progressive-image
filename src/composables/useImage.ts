import { ref, computed, isRef, nextTick } from "vue";
import { IMAGE_POLL_INTERVAL, IMAGE_ASPECT_RATIO } from "../constants";

export const useImage = (element) => {
  const image = new Image();
  const width = ref(0);
  const height = ref(0);

  const aspectRatio = computed(() => {
    return width.value ? height.value / width.value : IMAGE_ASPECT_RATIO;
  });

  const pollImageData = setInterval(() => {
    if (image && image.width) {
      clearInterval(pollImageData);

      width.value = image.width;
      height.value = image.height;
    }
  }, IMAGE_POLL_INTERVAL);

  const imageRenderer = (imageNode) => {
    const canvas = document.createElement("canvas");

    canvas.width = 1;
    canvas.height = 1;

    canvas.setAttribute("hidden", true);

    document.body.appendChild(canvas);

    canvas.getContext("2d").drawImage(imageNode, 0, 0);

    document.body.removeChild(canvas);
  };

  const loadImage = () => {
    const imageNode = isRef(element) ? element.value : element;
    const src = imageNode.src;

    image.src = src;

    if (image.complete) {
      return Promise.resolve();
    }

    return new Promise((resolve, reject) => {
      image.onload = () => {
        imageRenderer(imageNode);
        nextTick(resolve);
      };

      image.onerror = reject;
    });
  };

  return {
    width,
    height,
    aspectRatio,
    loadImage,
  };
};

export default useImage;
