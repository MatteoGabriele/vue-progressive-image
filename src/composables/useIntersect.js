import { INTERSECTION_THRESHOLD } from "@/constants";
import { isRef, nextTick, onMounted, onUnmounted, ref, watch } from "vue";

export const useIntersect = (element) => {
  const isIntersected = ref(false);
  const options = { threshold: INTERSECTION_THRESHOLD };
  const observer = new IntersectionObserver((entries) => {
    if (entries[0].isIntersecting) {
      isIntersected.value = true;
      observer.disconnect();
    }
  }, options);

  const watchIntersectionOnce = (callback) => {
    const stop = watch(
      isIntersected,
      (is) => {
        if (is) {
          nextTick().then(callback);
          stop();
        }
      },
      { immediate: true },
    );
  };

  onMounted(() => {
    const el = isRef(element) ? element.value : element;
    observer.observe(el);
  });

  onUnmounted(() => {
    observer.disconnect();
  });

  return {
    watchIntersectionOnce,
    isIntersected,
  };
};

export default useIntersect;
