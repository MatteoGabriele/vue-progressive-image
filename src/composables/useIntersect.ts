import { isRef, onMounted, onUnmounted, nextTick, ref, watch } from "vue";
import { INTERSECTION_THRESHOLD } from "@/constants";

export const useIntersect = (element) => {
  const isIntersected = ref(false);
  const options = { threshold: INTERSECTION_THRESHOLD };
  const observer = new IntersectionObserver((entries) => {
    if (entries[0].isIntersecting) {
      isIntersected.value = true;
      observer.disconnect();
    }
  }, options);

  const watchIntersectionOnce = (callback) => {
    const stop = watch(
      isIntersected,
      (is) => {
        if (is) {
          nextTick().then(callback);
          stop();
        }
      },
      { immediate: true }
    );
  };

  onMounted(() => {
    const el = isRef(element) ? element.value : element;
    observer.observe(el);
  });

  onUnmounted(() => {
    observer.disconnect();
  });

  return {
    watchIntersectionOnce,
    isIntersected,
  };
};

export default useIntersect;
