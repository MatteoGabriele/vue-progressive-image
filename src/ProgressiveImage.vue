<script setup lang="ts">
import useImage from "@/composables/useImage";
import useIntersect from "@/composables/useIntersect";
import {
  MAIN_IMAGE_LOAD_ERROR,
  MAIN_IMAGE_LOAD_SUCCESS,
  globalPropsKey,
} from "@/constants";
import type { ProgressiveImageProps } from "@/types";
import { computed, inject, onMounted, reactive, ref } from "vue";

const props = defineProps<ProgressiveImageProps>();
const globalProps = inject(globalPropsKey, {});
const mergedProps = reactive<ProgressiveImageProps>({
  ...props,
  ...globalProps,
});

const emit = defineEmits([MAIN_IMAGE_LOAD_SUCCESS, MAIN_IMAGE_LOAD_ERROR]);

const rootRef = ref(null);
const imageRef = ref(null);
const isMainImageRendered = ref(false);
const isFallbackImageRendered = ref(false);

const imageSrc = computed<string | undefined>(() => {
  return isFallbackImageRendered.value
    ? mergedProps.fallbackSrc
    : mergedProps.src;
});

const { isIntersected, watchIntersectionOnce } = useIntersect(rootRef);
const { loadImage, aspectRatio, width } = useImage(imageRef);
const isLoading = computed(() => !isMainImageRendered.value);

const paddingHack = computed(() => ({
  paddingBottom: `${aspectRatio.value * 100}%`,
}));

const componentStyle = computed(() => {
  if (mergedProps.objectCover || width.value === 0) {
    return;
  }

  return {
    maxWidth: `${width.value}px`,
  };
});

const imageClasses = computed(() => {
  return [
    mergedProps.customClass,
    {
      "v-progressive-image-object-cover": mergedProps.objectCover,
      "v-progressive-image-loading": isLoading.value,
    },
  ];
});

function toNumber(input: string | number | undefined): number {
  let value = 0;

  if (typeof input === "string") {
    value = Number.parseInt(input);
  } else if (typeof input === "number") {
    value = input;
  }

  return value;
}

function mainImageHandler() {
  loadImage()
    .then(() => {
      setTimeout(() => {
        isMainImageRendered.value = true;
        emit(MAIN_IMAGE_LOAD_SUCCESS);
      }, toNumber(mergedProps.delay));
    })
    .catch((error) => {
      isMainImageRendered.value = true;
      isFallbackImageRendered.value = true;
      emit(MAIN_IMAGE_LOAD_ERROR, error);
    });
}

onMounted(() => {
  if (mergedProps.placeholderSrc && mergedProps.blur) {
    document.documentElement.style.setProperty(
      "--progressive-image-blur",
      `${toNumber(mergedProps.blur)}px`,
    );
  }

  if (mergedProps.src) {
    watchIntersectionOnce(mainImageHandler);
  }
});
</script>

<template>
  <div
    ref="rootRef"
    class="v-progressive-image"
    :class="imageClasses"
    :style="componentStyle"
  >
    <div :style="paddingHack">
      <transition
        :css="!mergedProps.placeholderSrc"
        name="v-progressive-image-main-fade"
        appear
      >
        <img
          v-if="isIntersected"
          v-show="isMainImageRendered"
          ref="imageRef"
          class="v-progressive-image-main"
          :src="imageSrc"
          :alt="mergedProps.alt"
          :title="mergedProps.title"
        />
      </transition>

      <template v-if="mergedProps.placeholderSrc">
        <transition name="v-progressive-image-placeholder-fade" appear>
          <img
            v-if="isLoading"
            class="v-progressive-image-placeholder"
            :loading="mergedProps.lazyPlaceholder ? 'lazy' : 'eager'"
            :src="mergedProps.placeholderSrc"
          />
        </transition>
      </template>

      <div v-if="$slots.default" class="v-progressive-image-slot-default">
        <slot :isLoading="isLoading" />
      </div>
    </div>
  </div>
</template>
