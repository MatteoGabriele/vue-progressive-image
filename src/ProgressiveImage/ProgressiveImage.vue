<script setup>
import { ref, onMounted, computed } from "vue";
import { MAIN_IMAGE_LOAD_SUCCESS, MAIN_IMAGE_LOAD_ERROR } from "@/constants";
import useImage from "@/composables/useImage";
import useIntersect from "@/composables/useIntersect";

const emit = defineEmits([MAIN_IMAGE_LOAD_SUCCESS, MAIN_IMAGE_LOAD_ERROR]);

const props = defineProps({
  src: String,
  placeholderSrc: String,
  fallbackSrc: String,
  alt: String,
  title: String,
  customClass: String,
  blur: [Number, String],
  lazyPlaceholder: {
    type: Boolean,
    default: false,
  },
  delay: {
    type: [Number, String],
    default: 0,
  },
  objectCover: {
    type: Boolean,
    default: false,
  },
});

const rootRef = ref(null);
const imageRef = ref(null);
const isMainImageRendered = ref(false);
const isFallbackImageRendered = ref(false);

const { isIntersected, watchIntersectionOnce } = useIntersect(rootRef);
const { loadImage, aspectRatio, width } = useImage(imageRef);
const isLoading = computed(() => !isMainImageRendered.value);

const paddingHack = computed(() => ({
  paddingBottom: `${aspectRatio.value * 100}%`,
}));

const componentStyle = computed(() => {
  if (props.objectCover || width.value === 0) {
    return;
  }

  return {
    maxWidth: `${width.value}px`,
  };
});

const imageClasses = computed(() => {
  return [
    props.customClass,
    {
      "v-progressive-image-object-cover": props.objectCover,
      "v-progressive-image-loading": isLoading.value,
    },
  ];
});

const mainImageHandler = () => {
  loadImage()
    .then(() => {
      setTimeout(() => {
        isMainImageRendered.value = true;
        emit(MAIN_IMAGE_LOAD_SUCCESS);
      }, props.delay * 1);
    })
    .catch((error) => {
      isMainImageRendered.value = true;
      isFallbackImageRendered.value = true;
      emit(MAIN_IMAGE_LOAD_ERROR, error);
    });
};

onMounted(() => {
  if (props.placeholderSrc && props.blur) {
    document.documentElement.style.setProperty(
      "--progressive-image-blur",
      `${props.blur * 1}px`
    );
  }

  if (props.src) {
    watchIntersectionOnce(mainImageHandler);
  }
});
</script>

<template>
  <div
    ref="rootRef"
    class="v-progressive-image"
    :class="imageClasses"
    :style="componentStyle"
  >
    <div :style="paddingHack">
      <transition
        :css="!placeholderSrc"
        name="v-progressive-image-main-fade"
        appear
      >
        <img
          v-if="isIntersected"
          v-show="isMainImageRendered"
          ref="imageRef"
          class="v-progressive-image-main"
          :src="isFallbackImageRendered ? fallbackSrc : src"
          :alt="alt"
          :title="title"
        />
      </transition>

      <template v-if="placeholderSrc">
        <transition name="v-progressive-image-placeholder-fade" appear>
          <img
            v-if="isLoading"
            class="v-progressive-image-placeholder"
            :loading="lazyPlaceholder ? 'lazy' : 'eager'"
            :src="placeholderSrc"
          />
        </transition>
      </template>

      <div v-if="$slots.default" class="v-progressive-image-slot-default">
        <slot :isLoading="isLoading" />
      </div>
    </div>
  </div>
</template>
