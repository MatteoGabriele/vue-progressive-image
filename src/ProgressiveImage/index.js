import ProgressiveImage from "./ProgressiveImage.vue";

export { ProgressiveImage };

export default ProgressiveImage;
