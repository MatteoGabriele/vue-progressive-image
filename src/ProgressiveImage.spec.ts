import { ref, nextTick } from "vue";
import { flushPromises, mount } from "@vue/test-utils";
import type { Mock } from "vitest";

import ProgressiveImage from "@/ProgressiveImage.vue";
import { useIntersect } from "@/composables/useIntersect";
import { useImage } from "@/composables/useImage";

vi.mock("@/composables/useIntersect", async (importOriginal) => {
  const actual = await importOriginal();
  return {
    ...(typeof actual === "object" ? actual : {}),
    useIntersect: vi.fn(),
  };
});

vi.mock("@/composables/useImage", async (importOriginal) => {
  const actual = await importOriginal();
  return {
    ...(typeof actual === "object" ? actual : {}),
    useImage: vi.fn(),
  };
});

describe("ProgressiveImage", () => {
  beforeEach(() => {
    vi.useFakeTimers();

    (useIntersect as Mock).mockReturnValue({
      isIntersecting: ref(true),
      isReady: ref(true),
      hasIntersectedOnce: ref(true),
      watchIntersectionOnce: vi.fn((callback) => {
        callback();
      }),
    });

    (useImage as Mock).mockReturnValue({
      width: ref(700),
      height: ref(400),
      aspectRatio: ref(0.5625),
      loadImage: vi.fn(() => Promise.resolve()),
    });
  });

  afterEach(() => {
    vi.clearAllMocks();
    vi.useRealTimers();
  });

  test("render component", () => {
    const wrapper = mount(ProgressiveImage, {
      props: {
        src: "main-image.jpg",
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test("the image should be visible", async () => {
    const wrapper = mount(ProgressiveImage, {
      attachTo: document.body,
      props: {
        src: "main-image.jpg",
      },
    });

    await flushPromises(); // wait image load
    vi.runAllTimers(); // run delay timer
    await nextTick(); // wait next rendering cycle

    expect(wrapper.find("img.v-progressive-image-main").isVisible()).toEqual(
      true
    );
  });

  test("the image is not rendered when is not intersected", () => {
    (useIntersect as Mock).mockReturnValue({
      isIntersecting: ref(false),
      isReady: ref(true),
      hasIntersectedOnce: ref(false),
      watchIntersectionOnce: vi.fn(),
    });

    const wrapper = mount(ProgressiveImage, {
      props: {
        src: "main-image.jpg",
      },
    });

    expect(wrapper.find("img.v-progressive-image-main").exists()).toEqual(
      false
    );
  });

  test("render image with placeholder", () => {
    const wrapper = mount(ProgressiveImage, {
      props: {
        src: "main-image.jpg",
        placeholderSrc: "placeholder-image.jpg",
      },
    });

    expect(
      wrapper.find("img.v-progressive-image-placeholder").exists()
    ).toEqual(true);
  });

  test("use alt attribute", () => {
    const wrapper = mount(ProgressiveImage, {
      props: {
        src: "main-image.jpg",
        placeholderSrc: "placeholder-image.jpg",
        alt: "image description",
      },
    });

    expect(
      wrapper.find("img.v-progressive-image-main").attributes("alt")
    ).toEqual("image description");
  });

  test("use object cover", () => {
    const wrapper = mount(ProgressiveImage, {
      props: {
        src: "main-image.jpg",
        placeholderSrc: "placeholder-image.jpg",
        objectCover: true,
      },
    });

    expect(
      wrapper
        .find(".v-progressive-image.v-progressive-image-object-cover")
        .exists()
    ).toEqual(true);
  });

  test("render default slot", () => {
    const wrapper = mount(ProgressiveImage, {
      slots: {
        default: "<div>lorem ipsum</div>",
      },
      props: {
        src: "main-image.jpg",
      },
    });

    expect(wrapper.find(".v-progressive-image-slot-default").text()).toEqual(
      "lorem ipsum"
    );
  });

  test("lazy load placeholder images", () => {
    const wrapper = mount(ProgressiveImage, {
      props: {
        src: "main-image.jpg",
        placeholderSrc: "placeholder-image.jpg",
        lazyPlaceholder: true,
      },
    });

    expect(
      wrapper.find("img.v-progressive-image-placeholder").attributes("loading")
    ).toEqual("lazy");
  });

  test("render title attributes", () => {
    const wrapper = mount(ProgressiveImage, {
      props: {
        src: "main-image.jpg",
        title: "lorem ipsum dolor sit amet",
      },
    });

    expect(
      wrapper.find("img.v-progressive-image-main").attributes("title")
    ).toEqual("lorem ipsum dolor sit amet");
  });
});
