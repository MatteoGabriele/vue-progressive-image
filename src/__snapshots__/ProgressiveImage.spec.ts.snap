// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProgressiveImage > render component 1`] = `
<div
  class="v-progressive-image v-progressive-image-loading"
  style="max-width: 700px;"
>
  <div
    style="padding-bottom: 56.25%;"
  >
    <transition-stub
      appear="true"
      css="true"
      mode="out-in"
      name="v-progressive-image-main-fade"
      persisted="false"
    >
      <img
        class="v-progressive-image-main"
        decoding="async"
        height="400"
        loading="lazy"
        src="main-image.jpg"
        style="display: none;"
        width="700"
      />
    </transition-stub>
    <!--v-if-->
    <!--v-if-->
  </div>
</div>
`;
